### PR TITLE
using same regex for identifying verions for running migrations and schema loading

### PR DESCRIPTION
**Pull request by onomojo**
Currently rails migrations use a different regex than schema loading for identifying migrations to be run. See: https://github.com/rails/rails/blob/master/activerecord/lib/active_record/migration.rb#L1070

This pull request makes schema loading use the same regex so that loading the schema behaves the same as running migrations.

As a positive side effect, you can now organize your old database migrations into subdirectories without it completely breaking everything.

### CODE REVIEW

**Comment by maclover7** [comment]
Should we add a regression test here?

**Comment by onomojo** [comment]
I considered it but since the method already lacks any tests around it and the change is relatively trivial I didn't have the motivation.

**Comment by eileencodes** [comment]
Regardless of whether there are already tests in place or not, this needs a regression test.

**Comment by onomojo** [comment]
Sorry I'm just not interested in catching up with your test debt.

**Comment by onomojo** [comment]
I just want to point out that this has been untested for 8 years now with numerous changes made since that time: https://github.com/onomojo/rails/commit/8a5a9dcbf64843f064b6e8a0b9c6eea8f0b8536e#diff-a0775e1ec64264dc76a8ee71a5211ae3  I find it just a little humorous that after 8 years you now require tests on it especially when the change is so ridiculously trivial.

**Comment by onomojo** [comment]
There's a test

**Comment by onomojo** [comment]
The tests are passing locally. I'm not sure why the sqlite version is failing. The failure is what I'd expect if it was running the old regex which leads me to think Travis is using a cached version somehow in that instance. Ideas?

**Comment by onomojo** [comment]
I just tested this inside this VM as well (https://github.com/rails/rails-dev-box) and test pass fine:

```
vagrant@rails-dev-box:/vagrant/rails/activerecord$ bundle exec rake TEST=test/cases/schema_loading_test.rb 
/usr/bin/ruby2.3 -w -I"lib:test" -I"/var/lib/gems/2.3.0/gems/rake-11.1.2/lib" "/var/lib/gems/2.3.0/gems/rake-11.1.2/lib/rake/rake_test_loader.rb" "test/cases/schema_loading_test.rb" 
Using mysql2
Run options: --seed 13645

# Running:

.

Finished in 0.026866s, 37.2222 runs/s, 74.4445 assertions/s.

1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
/usr/bin/ruby2.3 -w -I"lib:test" -I"/var/lib/gems/2.3.0/gems/rake-11.1.2/lib" "/var/lib/gems/2.3.0/gems/rake-11.1.2/lib/rake/rake_test_loader.rb" "test/cases/schema_loading_test.rb" 
Using sqlite3
Run options: --seed 46181

# Running:

.

Finished in 0.046745s, 21.3926 runs/s, 42.7852 assertions/s.

1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
/usr/bin/ruby2.3 -w -I"lib:test" -I"/var/lib/gems/2.3.0/gems/rake-11.1.2/lib" "/var/lib/gems/2.3.0/gems/rake-11.1.2/lib/rake/rake_test_loader.rb" "test/cases/schema_loading_test.rb" 
Using postgresql
Run options: --seed 44253

# Running:

.

Finished in 0.025476s, 39.2530 runs/s, 78.5059 assertions/s.

1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by rafaelfranca** [comment]
Maybe it is different versions of sqlite3? I know travis use a really old version of it.

**Comment by onomojo** [comment]
I'm a little skeptical its an sqlite version issue but with travis anything is possible. The regex change really only effects parsing the directory structures for the migrations and doesn't actually change anything related to the insertions into the database. The code basically loops over the versions the regex finds and does insertions on each. Adding additional versions into that list shouldn't make any difference. It does insert 1 version, the version it'd find if it were using the old regex. I'm still thinking its testing against an old version of the code perhaps because of some caching considering the failure is exactly what would happen if tested against the old code.

I should also note that the entire test suite passes locally as well.

**Comment by onomojo** [comment]
Also regarding possible sqlite version issues, if that is the problem shouldn't the whole test suite be containerized so that kind of discrepancy doesn't happen on Travis? Having an official Rails dev VM that the CI server doesn't acctualy use is a bit of a half hearted effort at containerization.

**Comment by onomojo** [comment]
I decided to move the regex into a method instead of it being duplicated in both places. Migrations and schema loading needs to use the same regex so it seems to make more sense to call the same method instead of duping the regex itself in 2 places.

**Comment by onomojo** [comment]
@rafaelfranca Sure enough it was the sqlite3 version used on Travis. That version basically can't chain SQL commands together in the same execution. This: https://github.com/rails/rails/pull/24685/files  appeared to try to fix it but seems to have failed at least in this scenario. The latest commit ought to fix it.

**Comment by onomojo** [comment]
This is ready for review again.

**Comment by onomojo** [comment]
@matthewd Anything holding this up?

**Comment by bf4** [comment]
@onomojo you win finding old untested code.  I thought [since 2009 (6 years) was a long time :)](https://github.com/rails/rails/pull/22416)

**Comment by onomojo** [comment]
@maclover7 @matthewd  @eileencodes Anything holding this up from getting merged?

**Comment by shawjb** [comment]
looks like
```
rake db:migrate:status
```
is not able to find the migrations after moving migrations to a sub directories example output.
```
 Status   Migration ID    Migration Name
--------------------------------------------------
   up     20131223163427  ********** NO FILE **********
   up     20131223164106  ********** NO FILE **********
   up     20131223165228  ********** NO FILE **********
   up     20131223165637  ********** NO FILE **********
   up     20131223171925  ********** NO FILE **********
   up     20131223172837  ********** NO FILE **********
   up     20131223173116  ********** NO FILE **********
   up     20131223173747  ********** NO FILE **********
   up     20131223173919  ********** NO FILE **********
   up     20131223175417  ********** NO FILE **********
   up     20131223175837  ********** NO FILE **********
   up     20131223214613  ********** NO FILE **********
   up     20131223214616  ********** NO FILE **********
   up     20131227185736  ********** NO FILE **********
   up     20131227185838  ********** NO FILE **********
   up     20131227221544  ********** NO FILE **********
```
SNIP
```
   up     20151218180943  ********** NO FILE **********
   up     20151221141139  ********** NO FILE **********
   up     20151221142854  ********** NO FILE **********
   up     20160104145248  Add app to study plan states
   up     20160104154714  Non null constraint on purchase receipts
```

File: activerecord/test/cases/schema_loading_test.rb [hunk 1]
```diff
@@ -1,52 +1,41 @@
...
+    assert_equal 3, ActiveRecord::Migrator.current_version
+
+    sm_table = ActiveRecord::Migrator.schema_migrations_table_name
+    migrated = ActiveRecord::Base.connection.select_values("SELECT version FROM #{sm_table}").map(&:to_i)
+    assert_equal [1,2,3], migrated.sort
   end
 
-  private
+  if current_adapter?(:SQLite3Adapter)
+    %w{3.7.8 3.7.11 3.7.12}.each do |version_string|
```
Comment (rafaelfranca) [inline activerecord/test/cases/schema_loading_test.rb]: What is this doing? It doesn't seem to change the version of sqlite3 that is being tested.

Comment (onomojo) [inline activerecord/test/cases/schema_loading_test.rb]: I just mimicked what was being done in here:

https://github.com/rails/rails/blob/master/activerecord/test/cases/schema_dumper_test.rb

And assumed it was actually working.

Comment (rafaelfranca) [inline activerecord/test/cases/schema_loading_test.rb]: That test is stubbing the connection version too. https://github.com/rails/rails/blob/master/activerecord/test/cases/schema_dumper_test.rb#L35-L36

Comment (onomojo) [inline activerecord/test/cases/schema_loading_test.rb]: Good catch. I just added that too.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb [hunk 2]
```diff
@@ -991,16 +991,27 @@ def dump_schema_information #:nodoc:
...
         else
           versions.map { |version|
-            "INSERT INTO #{sm_table} (version) VALUES ('#{version}');"
-          }.join "\n\n"
+            "INSERT INTO #{sm_table} (version) VALUES ('#{version}');\n\n"
+          }.join ""
+        end
+      end
+
+      def execute_multi_insert(sql)
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: `# :nodoc:`

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb [hunk 3]
```diff
@@ -991,16 +991,27 @@ def dump_schema_information #:nodoc:
...
+        if supports_multi_insert?
+          execute sql
+        else
+          # executing separately because old versions of sqlite3 can't reliably execute multiple statements at once
+          statements = sql.split(';').select{|i| i.present? } 
+          statements.each do |statement|
+            execute "#{statement};"
```
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: No need string interpolation. `execute statement`.

Comment (onomojo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: It adds the ; back into the statement.

File: activerecord/lib/active_record/migration.rb [hunk 4]
```diff
@@ -1067,7 +1067,7 @@ def parse_migration_filename(filename) # :nodoc:
       def migrations(paths)
         paths = Array(paths)
 
-        files = Dir[*paths.map { |p| "#{p}/**/[0-9]*_*.rb" }]
+        files = Dir[*paths.map { |p| migration_paths_regex(p) }]
```
Comment (kamipo) [inline activerecord/lib/active_record/migration.rb]:
```ruby
      def migrations(paths)
        paths = Array(paths)
  
        files = migration_files(paths)
...(snip)...
      end

      def migration_files(paths)
        Dir[*paths.map { |p| "#{p}/**/[0-9]*_*.rb" }]
      end
```

How about it?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb (hunk 3 above)
Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: Why need the `;`?

File: activerecord/lib/active_record/migration.rb (hunk 4 above)
Comment (onomojo) [inline activerecord/lib/active_record/migration.rb]: That would work as well. My initial thinking was to make as little change as possible but it seems that's going to be impossible. I'm going to wait until I have the time to unravel the db:migrate:status issue also before I take a side. If we can reuse this method there as well, which I suspect we can, then I'm all for changing it.

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb (hunk 3 above)
Comment (onomojo) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: I dunno. Maybe its useless. I'll try removing it and see if it still works.

File: activerecord/lib/active_record/migration.rb (hunk 4 above)
Comment (kamipo) [inline activerecord/lib/active_record/migration.rb]: This code will fix `db:migrate:status `.

```diff
--- a/activerecord/lib/active_record/railties/databases.rake
+++ b/activerecord/lib/active_record/railties/databases.rake
@@ -112,9 +112,9 @@ db_namespace = namespace :db do
       end
       db_list = ActiveRecord::SchemaMigration.normalized_versions
 
+      migration_paths = ActiveRecord::Tasks::DatabaseTasks.migrations_paths
       file_list =
-          ActiveRecord::Tasks::DatabaseTasks.migrations_paths.flat_map do |path|
-            Dir.foreach(path).map do |file|
+            ActiveRecord::Migrator.migration_files(migration_paths).map do |file|
               next unless ActiveRecord::Migrator.match_to_migration_filename?(file)
 
               version, name, scope = ActiveRecord::Migrator.parse_migration_filename(file)
@@ -122,7 +122,6 @@ db_namespace = namespace :db do
               status = db_list.delete(version) ? 'up' : 'down'
               [status, version, (name + scope).humanize]
             end.compact
-          end
 
       db_list.map! do |version|
         ['up', version, '********** NO FILE **********']
```

Comment (onomojo) [inline activerecord/lib/active_record/migration.rb]: 👍

**Comment by onomojo** [comment]
Does anyone review pull requests on this project?

**Comment by bf4** [comment]
@onomojo so, you know the answer is yes because you've had multiple people reviewing it back in May.  I also have pending prs, some of which seems like no-brainers, but, seem to get lost in the shuffle.  But, I'm gonna guess there's more productive ways to ask for attention. :)  (Also, probably best to rebase when asking ;)

**Comment by kamipo** [comment]
Closing since it has been fixed in #28287.

**Pull request closed without merging**